File: wallet/src/main/java/dev/mistial/microcard/wallet/DeviceCbor.java
package dev.mistial.microcard.wallet;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;

/** Fixed-field device records; JSON remains a host authoring format. */
final class DeviceCbor {
    private DeviceCbor() {}

    static byte[] managementNames(String first, String second) {
        var output = new ByteArrayOutputStream(134);
        output.write(0x83);
        output.write(1);
        for (String value : new String[]{first, second}) {
            if (value == null || !value.matches("[A-Za-z0-9][A-Za-z0-9_.-]{0,63}"))
                throw new IllegalArgumentException("Invalid management identifier");
            byte[] bytes = value.getBytes(StandardCharsets.US_ASCII);
            if (bytes.length < 24) output.write(0x60 + bytes.length);
            else { output.write(0x78); output.write(bytes.length); }
            output.writeBytes(bytes);
        }
        return output.toByteArray();
    }
}

File: scripts/device_cbor.py
"""Versioned deterministic CBOR device records. Host reports may still use JSON."""
import re


def management_names(first, second):
    values = []
    for value in (first, second):
        if not isinstance(value, str) or re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9_.-]{0,63}", value) is None:
            raise ValueError("invalid management identifier")
        encoded = value.encode("ascii")
        length = len(encoded)
        values.append((bytes([0x60 + length]) if length < 24 else bytes([0x78, length])) + encoded)
    return b"\x83\x01" + b"".join(values)
